### PR TITLE
Strict check and efficient optimize ResumeUploader

### PR DESCRIPTION
**Pull request by netroby**

### CODE REVIEW

**Comment by longbai** [comment]
这速度真快，等下CI

**Comment by netroby** [comment]
I see, there were some more bugs, working on it

File: src/Qiniu/Storage/ResumeUploader.php [hunk 1]
```diff
@@ -74,18 +74,24 @@ public function upload()
...
+            if ((is_array($ret) && (!array_key_exists('crc32', $ret)
+                || (array_key_exists('crc32', $ret) && $crc !== $ret['crc32'])))
+                || $response->needRetry()
+            ) {
                 $response = $this->makeBlock($data, $blockSize);
                 $ret = $response->json();
             }
 
-            if (! $response->ok() || !isset($ret['crc32'])|| $crc != $ret['crc32']) {
+            if ((is_array($ret) && (!array_key_exists('crc32', $ret)
```
Comment (longbai) [inline src/Qiniu/Storage/ResumeUploader.php]: 这个地方逻辑错误

File: src/Qiniu/Storage/ResumeUploader.php [hunk 2]
```diff
@@ -74,18 +74,24 @@ public function upload()
...
             $ret = null;
-            if ($response->ok() && $response->json() != null) {
+            if ($response->ok() && $response->json() !== null) {
                 $ret = $response->json();
             }
             if ($response->statusCode < 0) {
                 $this->host = $this->config->getUpHostBackup();
             }
-            if ($response->needRetry() || !isset($ret['crc32']) || $crc != $ret['crc32']) {
+            if ((is_array($ret) && (!array_key_exists('crc32', $ret)
```
Comment (longbai) [inline src/Qiniu/Storage/ResumeUploader.php]: 这个地方短路判断，不需要再判断 array_key_exists('crc32')

**Comment by longbai** [comment]
@netroby 麻烦改下，多谢

**Comment by netroby** [comment]
I can't understand the after upload check, and the tests cases. so i do not know how to fix it.

**Comment by netroby** [comment]
Close it now, i have not idea how to deal with it.

**Pull request closed without merging**